src/aa_remove_data/pb_utils.py

```diff
@@ -0,0 +1,164 @@
+
+    def get_datastr(self, sample: type, year: int) -> str:
+        """Get a string contaiing information about a sample.
+        Args:
+            sample (type): A sample from a pb file.
+            year (int): The year the sample was collected.
+        Returns:
+            str: A string containing the sample information.
+        """
+        date = self.convert_to_datetime(year, sample.secondsintoyear)
+        return (
+            f"{date}    {sample.secondsintoyear:8d}    {sample.nano:9d}"
+            f"    {sample.val}\n"
+        )
+
+    def get_sample_type(self) -> str:
+        """Get the name of a pb file's sample type using information in its
+        header.
+
+        Returns:
+            str: Name of sample type, e.g VECTOR_DOUBLE.
+        """
+        type_descriptor = self.header.DESCRIPTOR.fields_by_name["type"]
+        enum_descriptor = type_descriptor.enum_type
+        return enum_descriptor.values_by_number[self.header.type].name
+
+    def get_sample_class(self) -> type:
+        """Get the EPICSEvent_pb2 class corresponding to samples in a pb file.
+        Instances of this class can interpret pb samples of a matching type.
+
+        Returns:
+            type: pb sample class.
+        """
+        # Ensure self.sample_type is set first.
+        if not self.sample_type:
+            self.sample_type = self.get_sample_type()
+        sample_type_camel = self._convert_to_class_name(self.sample_type)
+        sample_class = getattr(EPICSEvent_pb2, sample_type_camel)
+        return sample_class
+
+    def write_to_txt(self, filepath: PathLike):
+        """Write a text file from a PBUtils object.
+
+        Args:
+            filepath (PathLike): Filepath for file to be written.
+        """
+        pvname = self.header.pvname
+        year = self.header.year
+        sample_type = self.get_sample_type()
+        data_strs = [self.get_datastr(sample, year) for sample in self.samples]
+        with open(filepath, "w") as f:
+            f.write(f"{pvname}, {sample_type}, {year}\n")
+            f.write(f"DATE{' ' * 19}SECONDS{' ' * 5}NANO{' ' * 9}VAL\n")
+            f.writelines(data_strs)
+
+    def read_pb(self, filepath: PathLike):
+        """Read a pb file that is structured in the Archiver Appliance format.
+        Gathers the header and samples from this file and assigns them to
+        self.header self.samples.
+
+        Args:
+            filepath (PathLike): Path to pb file.
+        """
+        with open(filepath, "rb") as f:
+            first_line = self._unescape_data(f.readline().strip())
+            self.header.ParseFromString(first_line)
+            sample_class = self.get_sample_class()
+            lines = f.readlines()
+            self.samples = [sample_class() for n in range(len(lines))]
+            for i, sample in enumerate(self.samples):
+                line = self._unescape_data(lines[i].strip())
+                sample.ParseFromString(line)
```

Comment:
Similarly with this processing? Can the file be closed before the lines are processed? It feels better to keep only the read elements within the `with` block.

Reply:
Everything other than `readline` functions has been taken out of the `with open` block.